File: bnhsdk/ios/Classes/BinahEventChannel.swift
import Foundation


class BinahEventChannel: NSObject, FlutterStreamHandler {
    
    static let eventChannelId = "plugins.binah.ai/sdk_events"
    private var eventSink: FlutterEventSink?
    
    init(messenger: FlutterBinaryMessenger) {
        super.init()
        let channel = FlutterEventChannel(name: BinahEventChannel.eventChannelId, binaryMessenger: messenger)
        channel.setStreamHandler(self)
    }
    
    func sendEvent(name: String, payload: Any) {
        DispatchQueue.main.async {
            self.eventSink?([
                "event": name,
                "payload": payload
            ])
        }
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: bnhsdk/android/src/main/kotlin/ai/binah/binah_flutter_sdk/BinahEventChannel.kt
package ai.binah.binah_flutter_sdk

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class BinahEventChannel(
    messenger: BinaryMessenger,
    name: String
): EventChannel.StreamHandler {

    companion object {
        const val eventChannelId = "plugins.binah.ai/sdk_events"
    }

    private var eventSink: EventChannel.EventSink?  = null
    private val mainScope = CoroutineScope(Dispatchers.Main)

    init {
        val channel = EventChannel(messenger, name)
        channel.setStreamHandler(this)
    }

    fun sendEvent(name: String, payload: Any) {
        eventSink?.let { sink ->
            mainScope.launch {
                sink.success(mapOf(
                    Pair("event", name),
                    Pair("payload", payload)
                ))
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

}